plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -161,31 +236,34 @@ fu_genesys_usbhub_device_read_flash(FuGenesysUsbhubDevice *self,
 				    FuProgress *progress,
 				    GError **error)
 {
-	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(self));
 	g_autoptr(GPtrArray) chunks = NULL;
 
-	chunks = fu_chunk_array_mutable_new(buf, bufsz, start_addr, 0x0, self->flash_rw_size);
+	chunks = fu_chunk_array_mutable_new(buf,
+					    bufsz,
+					    start_addr,
+					    self->flash_block_size,
+					    self->flash_rw_size);
 	if (progress != NULL) {
 		fu_progress_set_id(progress, G_STRLOC);
 		fu_progress_set_steps(progress, chunks->len);
 	}
 	for (guint i = 0; i < chunks->len; i++) {
 		FuChunk *chk = g_ptr_array_index(chunks, i);
-
-		if (!g_usb_device_control_transfer(usb_device,
-						   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
-						   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
-						   G_USB_DEVICE_RECIPIENT_DEVICE,
-						   self->vcs.req_read,
-						   (fu_chunk_get_address(chk) & 0x0f0000) >>
-						       4,			       /* value */
-						   fu_chunk_get_address(chk) & 0xffff, /* idx */
-						   fu_chunk_get_data_out(chk),	       /* data */
-						   fu_chunk_get_data_sz(chk), /* data length */
-						   NULL,		      /* actual length */
-						   GENESYS_USBHUB_USB_TIMEOUT,
-						   NULL,
-						   error)) {
+		g_autofree FuGenesysUsbSetup *setup = g_malloc0(sizeof(FuGenesysUsbSetup));
+
+		setup->req_type.dir = G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST;
+		setup->req_type.type = G_USB_DEVICE_REQUEST_TYPE_VENDOR;
+		setup->req_type.recipient = G_USB_DEVICE_RECIPIENT_DEVICE;
+		setup->request = self->vcs.req_read;
+		setup->value = (fu_chunk_get_page(chk) & 0x000f) << 12;
+		setup->index = fu_chunk_get_address(chk) & 0xffff;
+		setup->length = fu_chunk_get_data_sz(chk);
```

Comment:
does using the USB names add clarity here, or make it worse? I'd argue a custom struct with setup->address = fu_chunk_get_address(chk) would be easier to understand -- i.e. you store them in the struct what they *are* not what you're sending them *as*

Reply:
> does using the USB names add clarity here, or make it worse?

I think this function already named the purpose, and `fu_genesys_usbhub_device_send_control_transfer` too. Therefore, add a middle-function to do all the R/W again, maybe cause some confuse?

Reply:
Yup, maybe it's confusing.